### ISSUE TITLE
Spurious warning in #[allow(clippy::non_send_fields_in_send_ty)] for complex thread-safe wrapper

### ISSUE THREAD

**Issue by alice-i-cecile** [rust-lang/rust-clippy#8275]
### Summary

Our unsafe implementation of `Send` for a wrapper type that deliberately strips out all of the `!Send` components of our complex collection tripped the new clippy lint. This was, in fact the entire point of the type, and why we used an `unsafe impl` block.

Ping @DJMcNab, the author of that PR.

### Lint Name

#[allow(clippy::non_send_fields_in_send_ty)]

### Reproducer

See https://github.com/bevyengine/bevy/pull/3519 for the case where this lint triggered spuriously.

The existing warning on `World` was correct: this impl was in fact unsound, and this PR was an attempt to solve that problem.

### Version

```text
rustc 1.60.0-nightly (1bd4fdc94 2022-01-12)
binary: rustc
commit-hash: 1bd4fdc943513e1004f498bbf289279c9784fc6f
commit-date: 2022-01-12
host: x86_64-pc-windows-msvc
release: 1.60.0-nightly
LLVM version: 13.0.0
```

**Comment by xFrednet** [comment]
Thank you for the report. The lint sadly has several false positives.

There will be a 1.58.1 release to make the lint allow by default on next Thursday. (See #8045)

**Comment by Qwaz** [comment]
Thanks for sharing the spurious failures. It's still good to hear that the lint was able to prevent a soundness bug. Hope we can bring this lint back to enable-by-default in the future after solving these known false positives.
